src/compiler/moduleNameResolver.ts

```diff
@@ -788,13 +788,16 @@ namespace ts {
     }
 
     function tryLoadModuleUsingPathsIfEligible(extensions: Extensions, moduleName: string, loader: ResolutionKindSpecificLoader, state: ModuleResolutionState) {
-        const { baseUrl, paths } = state.compilerOptions;
-        if (baseUrl && paths && !pathIsRelative(moduleName)) {
+        const { baseUrl, paths, pathsBasePath } = state.compilerOptions;
+        if (paths && !pathIsRelative(moduleName)) {
             if (state.traceEnabled) {
-                trace(state.host, Diagnostics.baseUrl_option_is_set_to_0_using_this_value_to_resolve_non_relative_module_name_1, baseUrl, moduleName);
+                if (baseUrl) {
+                    trace(state.host, Diagnostics.baseUrl_option_is_set_to_0_using_this_value_to_resolve_non_relative_module_name_1, baseUrl, moduleName);
+                }
                 trace(state.host, Diagnostics.paths_option_is_specified_looking_for_a_pattern_to_match_module_name_0, moduleName);
             }
-            return tryLoadModuleUsingPaths(extensions, moduleName, baseUrl, paths, loader, /*onlyRecordFailures*/ false, state);
+            const baseDirectory = baseUrl ?? Debug.checkDefined(pathsBasePath || state.host.getCurrentDirectory?.(), "Encountered 'paths' without a 'baseUrl', config file, or host 'getCurrentDirectory'.");

```

Comment:
This looks sketchy, but I manually verified that the assertion should hold in all cases we currently have; it’s just hard to encode it into the types. The only times `host.getCurrentDirectory` is `undefined` are

1. The command line parser and typings installer each call the module name resolver in an ad-hoc way under certain circumstances, but they pass a hard-coded `compilerOptions` that lack `paths`, so this code path is never taken.
2. `Project` calls `getAutomaticTypeDirectives` with a `DirectoryStructureHost`, but that function never uses `host.getCurrentDirectory()`.

Changing `ModuleResolutionHost['getCurrentDirectory']` to be non-optional makes these call sites a bit of a nuisance.